### PR TITLE
mptcp: redundant: new scheduler mptcp_redundant.c

### PR DESCRIPTION
**Pull request by ghost**
The redundant scheduler sends data replicated through all the available and
active subflows. Backup subflows are only used in case of data retransmissions.

### CODE REVIEW

File: net/mptcp/Kconfig [hunk 1]
```diff
@@ -85,6 +85,15 @@ config MPTCP_ROUNDROBIN
 	  This is a very simple round-robin scheduler. Probably has bad performance
 	  but might be interesting for researchers.
 
+config MPTCP_REDUNDANT
+	tristate "MPTCP Redundant"
+	depends on (MPTCP=y)
```
Comment (cpaasch) [inline net/mptcp/Kconfig:90]: We probably should make this depend on MPTCP_SCHED_ADVANCED (same change is needed for MPTCP_ROUNDROBIN). But that would be a separate pull-request (as MPTCP_ROUNDROBIN needs to change as well).
Feel free to make one, if you want to.

File: net/mptcp/mptcp_redundant.c [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version
+ * 2 of the License, or (at your option) any later version.
+ */
+
+#include <linux/module.h>	/* Needed by all modules */
+#include <net/mptcp.h>
+
+/* Same as mptcp_dss_len from mptcp_output.c */
```
Comment (cpaasch) [inline net/mptcp/mptcp_redundant.c]: Then, export mptcp_dss_len out of mptcp_output.c. Let's not duplicate code.

**Comment by cpaasch** [comment]
As a general comment, there is way too much code-duplication. Please reduce this, because otherwise it is a nightmare to maintain it.

Thanks for posting the patches and looking forward to your updates!

**Comment by ghost** [comment]
I am not a C/kernel developer, so correct me if I am wrong.

As fas as I know, I think that static functions and global variables can only be "seen" in the file they were declared in. So I cannot reuse or "include" them from net/mptcp/mptcp_sched.c and net/mptcp/mptcp_output.c in my C file (net/mptcp/mptcp_redundant.c).

If that would be possible it would be great because this scheduler is the default one softly modified to do replication through active paths.

Thanks Christoph!

**Comment by cpaasch** [comment]
On 25/09/15 - 00:57:55, chrpinedo wrote:

> I am not a C/kernel developer, so correct me if I am wrong.
> 
> As fas as I know, I think that static functions and global variables can only be "seen" in the file they were declared in. So I cannot reuse or "include" them from net/mptcp/mptcp_sched.c and net/mptcp/mptcp_output.c in my C file (net/mptcp/mptcp_redundant.c).
> 
> If that would be possible it would be great because this scheduler is the default one softly modified to do replication through active paths.

You just have to remove the "static" from the functions in mptcp_sched.c and
export the function's declaration to a header function.

Christoph

**Comment by cpaasch** [comment]
Overall, I'm a bit concerned of sending segments to the reinject-queue and then pulling them out of there. Because, this means you are skb_copy_clone'ing plenty of segments over to the reinject-queue only to delete them afterwards as a subflow may not have any space in the congestion-window.

I think a better model would be to simply pick always the same segment in the meta-socket's send-queue . Then, you iterate over the subflows to find the next one (updating the path_mask of this segment). Each time you "test" as subflow, you do all the checks of whether this has already been sent on the subflow or not and whether there is space in the congestion-window. And you continue like this until you are at the end of the subflow list, upon which you proceed to the next segment in the meta-level send-queue.

I think this would be a simpler and cleaner approach.

**Comment by ghost** [comment]
As a summary, I have included all your concerns related to the code in the pull request (de-duplicate code, remove pr_debug functions, ...).

Regarding your proposal about the general approach for the module, it implies a new design and implementation of the module. It would require time to develop and to test in order to assure that the scheduler works as expected and as now. For example, at this time I don't know if this new implementation could be only restricted to the scheduler or if it is not self-contained and implies to modify other core functions of MPTCP.

**Comment by cpaasch** [comment]
Thanks for addressing my suggestions.

I think it would be beneficial to change the design for several reasons. First, it would allow to fully handle the case of new subflows popping up and unacked data sitting in the meta's send-queue. Further, putting segments in the reinject-queue is also prone to errors, because the reinject-queue assumes that the segments are in-order in this queue (see __mptcp_reinject_data()). Changing this might result in bugs in the processing in the reinject-queue (I'm concerned that the iterations in __mptcp_reinject_data() might have issues with out-of-order segments).

It is ok if core functions of MPTCP need to be adapted for the redundant scheduler, as long as those changes are clean and have a limited impact.

**Comment by ghost** [comment]
Ok I will try to address a new design for the scheduler. For that, do you have a document-reference-recomended_book to better understand the design of the MPTCP code (I suppose heavily based on TCP) or the only way is to read the code?

**Comment by cpaasch** [comment]
You can have a look at my thesis [1](chapter 5). I think as of today it's the most accurate description of the code. Otherwise, you can always ask questions on the list :).

http://inl.info.ucl.ac.be/publications/improving-multipath-tcp

**Comment by ghost** [comment]
I close this pull request because I feel that the initial code to include a redundant scheduler is provided in https://github.com/multipath-tcp/mptcp/pull/95 by @AlexanderFroemmgen.

**Pull request closed without merging**